File: BOBAPlaybook/Models/Showcase.swift
import Foundation

// MARK: - Showcase
//
// A named subset of the catalog — "WOBA" (Women of BOBA), each sport
// (Basketball / Football / Baseball / …), and eventually team-specific
// or user-defined curations. Used in three places:
//
//   • Find tab filter sheet — "Showcase" section under Card Type
//   • Learn > Browse tab — the same set of chips already lived here
//   • Search bar smart-match — typing "WOBA" or "Baseball" narrows to
//     the matching showcase without opening the filter sheet
//
// Keep all showcase definitions in this file so the three surfaces
// stay in sync without per-view duplication.

struct Showcase: Identifiable, Equatable, Sendable {
    let id: String
    let name: String
    /// Lowercase tokens that should resolve to this showcase when
    /// typed in the search bar. e.g. ["woba", "women of boba"].
    let searchTokens: [String]
    /// Short line shown beneath the chip label when hovered / in the
    /// Learn > Browse section.
    let description: String
    /// Approximate card count used for chip subtitles. Kept as a string
    /// since several showcases are compound sets whose exact count
    /// shifts as the catalog grows.
    let countLabel: String
    let match: @Sendable (Card) -> Bool

    static func == (l: Showcase, r: Showcase) -> Bool { l.id == r.id }
}

enum Showcases {

    // MARK: - WOBA

    /// Women of BOBA — every hero inspired by one of the 17 canonical
    /// women-in-sport athletes. Sourced from DISCORD_TERMINOLOGY.md §3
    /// + the roster posted to the trade channel.
    ///
    /// `nonisolated` on every static here because Swift 6 infers
    /// MainActor isolation on type-level statics in this module, and
    /// the `Showcase.match` closure is `@Sendable` — capturing a
    /// MainActor-isolated Set<String> inside a Sendable closure trips
    /// the compiler. Set<String> is inherently Sendable (immutable
    /// value type), so opting out of isolation is safe.
    nonisolated static let wobaHeroes: Set<String> = [
        "AJax", "Belladonna", "Brandi", "C.C.", "Cameleon", "Cheryl Bomb",
        "Coopanova", "Eraser", "Halo", "JPEG", "Lady Magic", "Leducky",
        "PB Buckets", "Pauldron", "Peek-A-Boo", "Ramponage", "Swoopes",
    ]

    nonisolated static let woba = Showcase(
        id: "woba",
        name: "WOMEN OF BOBA (WOBA)",
        searchTokens: ["woba", "women of boba", "women"],
        description: "Women of BOBA — heroes inspired by legendary female athletes across every sport.",
        countLabel: "884 cards"
    ) { card in
        Showcases.wobaHeroes.contains(card.hero)
    }

    // MARK: - Sports

    /// Sport → canonical athleteInspiration list. Keep the lists
    /// permissive — catalog spelling variants (McCaffrey/McCaffery,
    /// Tatis/Tatís) are already normalized in cards.json but we keep
    /// both spellings here in case a rogue variant slips through.
    nonisolated static let sportAthletes: [(label: String, athletes: Set<String>)] = [
        ("Basketball", ["LeBron James","Lebron James","Steph Curry","Kevin Durant",
                        "Giannis Antetokounmpo","Giannis Anteokounmpo","Giannis Antetetokounmpo",
                        "Nikola Jokic","Luka Doncic","Cooper Flagg","Paige Bueckers",
                        "Julius Erving","Magic Johnson","Allen Iverson","Kawhi Leonard",
                        "Ja Morant","Jayson Tatum","Jason Tatum","Caitlin Clark",
                        "Angel Reese","A'ja Wilson","Cynthia Cooper","Cheryl Miller",
                        "Sheryl Swoopes","Nancy Lieberman","Elena Delle Donne",
                        "Devin Booker","Anthony Edwards","Damian Lillard"]),
        ("Football",   ["Patrick Mahomes","Josh Allen","Lamar Jackson","Travis Kelce",
                        "Bo Jackson","Barry Sanders","Adrian Peterson","Derrick Henry",
                        "Justin Jefferson","Joe Burrow","Justin Herbert","CJ Stroud",
                        "Caleb Williams","Jordan Love","Aaron Rodgers","Saquan Barkley",
                        "Christian McCaffrey","Christian McCaffery","Tyreek Hill",
                        "Travis Hunter","Dak Prescott","Jalen Hurts","Jayden Daniels"]),
        ("Baseball",   ["Ken Griffey Jr.","Ken Griffey Sr.","Shohei Ohtani","Aaron Judge",
                        "Mike Trout","Mookie Betts","Juan Soto","Bo Jackson",
                        "Ronald Acuna Jr.","Fernando Tatis Jr.","Fernando Tatís Jr.",
                        "Vladimir Guerrero Jr.","Julio Rodriguez","Jackson Holliday",
                        "Paul Skenes","Rafael Devers","Bobby Witt Jr","Bobby Witt Jr.",
                        "Elly De La Cruz","Gunnar Henderson","Corbin Carroll","Jackson Chourio"]),
        ("Hockey",     ["Sidney Crosby","Alexander Ovechkin","Henrik Lundqvist"]),
        ("Tennis",     ["Jessica Pegula","Paula Badosa"]),
        ("Golf",       ["Bryson DeChambeau","Jordan Spieth"]),
        ("Soccer",     ["Brandi Chastain","Chastain","Christie Pearce Rampone","Jozy Altidore"]),
    ]

    /// Materialized Showcase list for every sport above. Generated once
    /// and cached so the three consuming surfaces share the same
    /// instances.
    nonisolated static let sports: [Showcase] = sportAthletes.map { entry in
        let athletes = entry.athletes
        return Showcase(
            id: "sport_\(entry.label.lowercased())",
            name: entry.label,
            searchTokens: [entry.label.lowercased()],
            description: "\(entry.label) athletes and their inspired heroes.",
            countLabel: ""
        ) { card in
            guard let insp = card.athleteInspiration else { return false }
            return athletes.contains(insp)
        }
    }

    // MARK: - Catalog

    /// All showcases visible as filter chips / searchable terms, in the
    /// order they should appear. New community or team-specific
    /// showcases (Cardinals, Red Sox, etc.) slot in here.
    nonisolated static let all: [Showcase] = [woba] + sports

    /// Look up a showcase by its id.
    static func byId(_ id: String) -> Showcase? {
        all.first { $0.id == id }
    }

    /// Resolve a raw search term (lowercase) to a showcase — used by the
    /// smart search path.
    static func matching(searchToken: String) -> Showcase? {
        let needle = searchToken.trimmingCharacters(in: .whitespaces).lowercased()
        guard !needle.isEmpty else { return nil }
        return all.first { $0.searchTokens.contains(needle) }
    }
}

File: android/core/domain/src/main/java/com/bobaplaybook/core/domain/showcase/Showcase.kt
package com.bobaplaybook.core.domain.showcase

import com.bobaplaybook.core.domain.model.Card

/**
 * Showcase — a named subset of the catalog. Mirrors iOS
 * `BOBAPlaybook/Models/Showcase.swift` verbatim so the cross-
 * platform behavior is identical.
 *
 * Three call sites:
 *  - Find FilterSheet "Showcase" picker
 *  - Find featured-ribbons no-search state
 *  - Search-bar smart-match (typing "WoBA" → narrows automatically)
 */
data class Showcase(
    val id: String,
    val name: String,
    /** Lowercase tokens that resolve to this showcase from the search bar. */
    val searchTokens: List<String>,
    val description: String,
    val countLabel: String,
    val match: (Card) -> Boolean,
)

object Showcases {

    // ─── WoBA — Women of BoBA ──────────────────────────────────
    val wobaHeroes: Set<String> = setOf(
        "AJax", "Belladonna", "Brandi", "C.C.", "Cameleon", "Cheryl Bomb",
        "Coopanova", "Eraser", "Halo", "JPEG", "Lady Magic", "Leducky",
        "PB Buckets", "Pauldron", "Peek-A-Boo", "Ramponage", "Swoopes",
    )

    val woba = Showcase(
        id = "woba",
        name = "WoBA (Women of BoBA)",
        searchTokens = listOf("woba", "women of boba", "women"),
        description = "Women of BOBA — heroes inspired by legendary female athletes across every sport.",
        countLabel = "884 cards",
    ) { card -> card.hero in wobaHeroes }

    // ─── Rookie Inspired ───────────────────────────────────────
    val rookieInspired = Showcase(
        id = "rookie_inspired",
        name = "Rookie Inspired",
        searchTokens = listOf("rookie", "rookie inspired", "rookies"),
        description = "Cards whose hero is inspired by an athlete in their rookie season at print time.",
        countLabel = "2,733 cards",
    ) { card -> card.rookieInspired }

    // ─── BoBA-team-curated Featured Collections ────────────────
    // Mirrors iOS BrowseFeaturedData.collections (Bo Jackson, Ken Griffey
    // Jr., Dr. J). Each is a hand-picked filter the design team flagged
    // as "worth surfacing on the explore tab."
    val bojax = Showcase(
        id = "bojax",
        name = "Bo Jackson",
        searchTokens = listOf("bo jackson", "bojax"),
        description = "The man who inspired it all — every BoJax card across every set and treatment.",
        countLabel = "147 cards",
    ) { card ->
        card.athleteInspiration == "Bo Jackson" ||
            card.hero == "BoJax" || card.hero == "Bojax"
    }

    val kenGriffeyJr = Showcase(
        id = "kid",
        name = "Ken Griffey Jr.",
        searchTokens = listOf("ken griffey", "griffey", "the kid"),
        description = "The Kid — one of baseball's most beloved players.",
        countLabel = "~76 cards",
    ) { card ->
        card.athleteInspiration == "Ken Griffey Jr." || card.hero == "The Kid"
    }

    val drJ = Showcase(
        id = "drj",
        name = "Dr. J",
        searchTokens = listOf("dr j", "dr. j", "julius erving"),
        description = "Julius Erving — the original aerial artist of basketball.",
        countLabel = "70 cards",
    ) { card ->
        card.athleteInspiration == "Julius Erving" || card.hero == "Dr. J"
    }

    // ─── Sport showcases ───────────────────────────────────────
    private val basketball = setOf(
        "LeBron James", "Lebron James", "Steph Curry", "Kevin Durant",
        "Giannis Antetokounmpo", "Giannis Anteokounmpo", "Giannis Antetetokounmpo",
        "Nikola Jokic", "Luka Doncic", "Cooper Flagg", "Paige Bueckers",
        "Julius Erving", "Magic Johnson", "Allen Iverson", "Kawhi Leonard",
        "Ja Morant", "Jayson Tatum", "Jason Tatum", "Caitlin Clark",
        "Angel Reese", "A'ja Wilson", "Cynthia Cooper", "Cheryl Miller",
        "Sheryl Swoopes", "Nancy Lieberman", "Elena Delle Donne",
        "Devin Booker", "Anthony Edwards", "Damian Lillard",
    )
    private val football = setOf(
        "Patrick Mahomes", "Josh Allen", "Lamar Jackson", "Travis Kelce",
        "Bo Jackson", "Barry Sanders", "Adrian Peterson", "Derrick Henry",
        "Justin Jefferson", "Joe Burrow", "Justin Herbert", "CJ Stroud",
        "Caleb Williams", "Jordan Love", "Aaron Rodgers", "Saquan Barkley",
        "Christian McCaffrey", "Christian McCaffery", "Tyreek Hill",
        "Travis Hunter", "Dak Prescott", "Jalen Hurts", "Jayden Daniels",
    )
    private val baseball = setOf(
        "Ken Griffey Jr.", "Ken Griffey Sr.", "Shohei Ohtani", "Aaron Judge",
        "Mike Trout", "Mookie Betts", "Juan Soto", "Bo Jackson",
        "Ronald Acuna Jr.", "Fernando Tatis Jr.", "Fernando Tatís Jr.",
        "Vladimir Guerrero Jr.", "Julio Rodriguez", "Jackson Holliday",
        "Paul Skenes", "Rafael Devers", "Bobby Witt Jr", "Bobby Witt Jr.",
        "Elly De La Cruz", "Gunnar Henderson", "Corbin Carroll", "Jackson Chourio",
    )
    private val hockey = setOf(
        "Connor McDavid", "Nathan MacKinnon", "Sidney Crosby", "Alexander Ovechkin",
        "Wayne Gretzky", "Mario Lemieux", "Auston Matthews", "Henrik Lundqvist",
    )
    private val soccer = setOf(
        "Lionel Messi", "Cristiano Ronaldo", "Kylian Mbappe", "Erling Haaland",
        "Brandi Chastain", "Chastain", "Christie Pearce Rampone", "Jozy Altidore",
    )
    private val boxing = setOf(
        "Floyd Mayweather Jr.", "Muhammad Ali", "Manny Pacquiao", "Sugar Ray Robinson",
    )
    // Smaller iOS-parity sports — added 2026-05-19 to match
    // BrowseFeaturedData.sports verbatim.
    private val tennis = setOf("Jessica Pegula", "Paula Badosa")
    private val golf = setOf("Bryson DeChambeau", "Jordan Spieth")

    val sportAthletes: List<Pair<String, Set<String>>> = listOf(
        "Basketball" to basketball,
        "Football"   to football,
        "Baseball"   to baseball,
        "Hockey"     to hockey,
        "Soccer"     to soccer,
        "Boxing"     to boxing,
        "Tennis"     to tennis,
        "Golf"       to golf,
    )

    val sports: List<Showcase> = sportAthletes.map { (sport, athletes) ->
        Showcase(
            id = "sport_${sport.lowercase()}",
            name = sport,
            searchTokens = listOf(sport.lowercase()),
            description = "$sport heroes — players whose card art is inspired by athletes in $sport.",
            countLabel = "—",
        ) { card -> card.athleteInspiration?.let { it in athletes } ?: false }
    }

    val all: List<Showcase> = listOf(
        woba, bojax, kenGriffeyJr, drJ, rookieInspired,
    ) + sports

    fun byId(id: String): Showcase? = all.firstOrNull { it.id == id }

    fun matching(searchToken: String): Showcase? {
        val needle = searchToken.lowercase().trim()
        if (needle.isEmpty()) return null
        return all.firstOrNull { showcase ->
            showcase.searchTokens.any { it == needle }
        }
    }
}
